### ISSUE TITLE
Execute composer config http-basic before run main script

### ISSUE THREAD

**Issue by elvishp2006** [renovatebot/config-help#491]
I got some private packages (repositories) on GitLab which I request them on my composer.json as a custom repository, when renovate bot ran he don't have access to those repositories, I need to execute the following command before main script of the docker image: `composer config http-basic.hq.apiki.com gitlab-ci-token $CI_JOB_TOKEN` there is a way to do that?

**Comment by rarkins** [comment]
No. But we pass authentication through to `composer` using `auth.json` so it could potentially be done that way. Source: https://github.com/renovatebot/renovate/blob/master/lib/manager/composer/artifacts.ts

**Comment by elvishp2006** [comment]
This local auth.json file https://github.com/renovatebot/renovate/blob/master/lib/manager/composer/artifacts.ts#L95 I can override him when running the docker image? If possible, how I can achive that?

**Comment by rarkins** [comment]
Before we continue, please fill out the issue template and edit your issue description above: https://github.com/renovatebot/config-help/issues/new?labels=&template=config-help.md&title=

**Comment by elvishp2006** [comment]
tks @rarkins updated the description.

**Comment by rarkins** [comment]
The `auth.json` file is *written out* by Renovate prior to calling `composer`, so it is not intended that you be able to overwrite the file. Instead, we are looking for a way for Renovate to craft the correct `auth.json` file contents such that `composer` knows how to authenticate with your server.

Looking at the [source](https://github.com/renovatebot/renovate/blob/fc0b3d3026bd823c1ec81bc2c9c35a79aec1f6b6/lib/manager/composer/artifacts.ts#L41-L84), the `auth.json` is constructed from `hostRules` using a combination of:
- any github.com or gitlab.com auth it finds
- any `packagist` hosts it finds, using http-basic authentication

Therefore the next question is: is this the exact same endpoint and token that you are running Renovate against?

**Comment by elvishp2006** [comment]
Yes, I'm running the renovate as a GitLab job as well:

```yaml
image: docker:19

services:
  - docker:dind

renovate:
  stage: build
  tags:
    - SharedRUNNERS-3
  script:
    - docker run --rm -e RENOVATE_TOKEN="$RENOVATE_TOKEN" renovate/renovate $(cat repositories.txt | xargs) --log-level=info --endpoint=https://hq.apiki.com/api/v4/ --platform=gitlab --token=$GITLAB_TOKEN
  only:
    - master
```

And that's my composer.json:

```json
...
  "repositories": [
    {
      "type": "vcs",
      "url": "https://hq.apiki.com/plugins-pagos/gravityforms.git"
    }
  ],
  "require": {
    "plugins-pagos/gravityforms": "2.4.15.15"
  }
...
```
### Artifact update problem

```
  [RuntimeException]                                                                                                                                                                         
  Failed to execute git clone --mirror 'https://hq.apiki.com/plugins-pagos/gravityforms.git' '/tmp/renovate/cache/others/composer/vcs/https---hq.apiki.com-plugins-pagos-gravityforms.git/'  
                                                                                                                                                                                             
  Cloning into bare repository '/tmp/renovate/cache/others/composer/vcs/https---hq.apiki.com-plugins-pagos-gravityforms.git'...                                                              
  remote: HTTP Basic: Access denied                                                                                                                                                          
  fatal: Authentication failed for 'https://hq.apiki.com/plugins-pagos/gravityforms.git/'  
```

**Comment by rarkins** [comment]
Are `GITLAB_TOKEN` and `RENOVATE_TOKEN` the same value? If so, you should be able to simplify your command by using only one. It won't make a difference for this particular Composer problem though, but I want to check this. I'm also interested to know if your tokens are dynamic (e.g. injected into env by the CI process) or if they're static (e.g. generated Personal Access Tokens).

**Comment by rarkins** [comment]
One more thing: is there an env variable that's set that would let us detect it's running on GitLab CI?

**Comment by elvishp2006** [comment]
> Are `GITLAB_TOKEN` and `RENOVATE_TOKEN` the same value? If so, you should be able to simplify your command by using only one. It won't make a difference for this particular Composer problem though, but I want to check this. I'm also interested to know if your tokens are dynamic (e.g. injected into env by the CI process) or if they're static (e.g. generated Personal Access Tokens).

The `GITLAB_TOKEN` and `RENOVATE_TOKEN` have different values, they are CI variables previously configurated on project setting.



> One more thing: is there an env variable that's set that would let us detect it's running on GitLab CI?

The variable `$CI_JOB_TOKEN` is a dynamical, one generated on CI process.

**Comment by rarkins** [comment]
`RENOVATE_TOKEN` and `--token=` are different ways to define the same thing. The CLI `--token=` will take precedence. I suggest first removing the env variable and verifying that things still work as before. Are they both personal access tokens or something else?

Then finally can you confirm that `CI_JOB_TOKEN` has the same rights a `RENOVATE_TOKEN`? i.e. they're essentially for authenticating with the same GitLab server?

**Comment by elvishp2006** [comment]
I did removed the `--token=` and It's working as before, and now I'm using just a personal access token. And yeah, I confirmed that `CI_JOB_TOKEN` got the same rights as a personal access token, I tested that command `composer config http-basic.hq.apiki.com gitlab-ci-token personal-access-token` work as well on my pc.
